### ISSUE TITLE
Service DLL not working in Windows 10

### ISSUE THREAD

**Issue by oalieno** [kevoreilly/CAPEv2#1476]
## About accounts on [capesandbox.com](https://capesandbox.com/)
* Issues isn't the way to ask for account activation. Ping capesandbox in [Twitter](https://twitter.com/capesandbox) with your username

## This is open source and you are getting __free__ support so be friendly!

# Prerequisites

Please answer the following questions for yourself before submitting an issue.

- [x] I am running the latest version
- [x] I did read the README!
- [x] I checked the documentation and found no answer
- [x] I checked to make sure that this issue has not already been filed
- [x] I'm reporting the issue to the correct repository (for multi-repository projects)
- [x] I have read and checked all configs (with all optional parts)
- [x] I'm aware that configs now are in `custom/conf/`

# Expected Behavior

I want to run service dll, but it failed.

# Current Behavior

I write a simple service dll that will run calc.exe in user session. If the service dll successfully executed, calc.exe will pop up. But it failed to execute the service dll. 

# Failure Information (for bugs)

I have investigated this issue myself and managed to fix it. However, I am not sure how to merge the solution upstream, and it is also related to capemon. My test sample is a 64-bit service DLL, so I haven't handled the 32-bit case. Additionally, I don't know how to run a 32-bit service DLL in Windows 10. I have tried several times but failed to start a 32-bit service DLL in Windows 10.

## TL;DR

I managed to fix some code and make 64-bit service dll works in Windows 10. Here is my fix
1. Fix WoW64 redirection, since our python is 32-bit program.
2. Fix capemon bugs.
    - inject mode parameter not working in capemon loader (can't specify inject mode)
    - unknown bugs in `CreateThreadBreakpoints`

## Fix WoW64 redirection

Since our python.exe is a 32-bit program (as the documentation suggests), there will be WoW64 redirection for all operations in python.exe, including file moving and registry setting. Regardless of whether the submitted samples are 32-bit or 64-bit, they will all be placed under C:\Windows\SysWOW64, and the registry will be in WOW6432NODE. If the submitted sample is a 32-bit service DLL and all the settings are for 32-bit, it should work. However, as I mentioned earlier, I couldn't get the service DLL to run successfully in Windows 10, even when testing on my machine. Therefore, I decided to focus on fixing the 64-bit service DLL issue. I tried to disable all the WoW64 redirection in our Python script, which includes the following three steps:

1. Disable WoW64 in https://github.com/kevoreilly/CAPEv2/blob/4d65838dfed81065c718bb95c68aaba029b06ae3/analyzer/windows/lib/common/abstracts.py#L166
4. Disable WoW64 in https://github.com/kevoreilly/CAPEv2/blob/4d65838dfed81065c718bb95c68aaba029b06ae3/analyzer/windows/modules/packages/service_dll.py#L93
5. In the following two lines, change `KEY_ALL_ACCESS` to `KEY_ALL_ACCESS | KEY_WOW64_64KEY` https://github.com/kevoreilly/CAPEv2/blob/4d65838dfed81065c718bb95c68aaba029b06ae3/analyzer/windows/modules/packages/service_dll.py#L76 https://github.com/kevoreilly/CAPEv2/blob/4d65838dfed81065c718bb95c68aaba029b06ae3/analyzer/windows/modules/packages/service_dll.py#L86
6. Delete the WOW64 value created by CreateServiceA, which will make the Service Control Manager (SCM) attempt to run the service in a 32-bit manner. See the code below.

I have provided the modified code in analyzer/windows/modules/packages/service_dll.py and analyzer/windows/lib/common/abstracts.py below, as well as the Python decorator I used to disable WoW64 redirection.

```python
...

def set_keys(self, servicename, dllpath):
        svchost_path = r"Software\Microsoft\Windows NT\CurrentVersion\Svchost"
        service_path = rf"System\CurrentControlSet\Services\{servicename}"
        parameter_path = rf"System\CurrentControlSet\Services\{servicename}\Parameters"

        try:
            log.info("Adding Parameters value: %s -> ServiceDll = %s", parameter_path, dllpath)
            with CreateKeyEx(HKEY_LOCAL_MACHINE, parameter_path, 0, KEY_ALL_ACCESS | KEY_WOW64_64KEY) as key:
                SetValueEx(key, "ServiceDll", 0, REG_EXPAND_SZ, dllpath)
        except Exception as e:
            log.info("Error setting registry value: %s", e)
            return

        # try to remove the WOW64 field in service registry, which is created by CreateServiceA
        try:
            with OpenKeyEx(HKEY_LOCAL_MACHINE, service_path, 0, KEY_ALL_ACCESS | KEY_WOW64_64KEY) as key:
                DeleteValue(key, "WOW64")
        except Exception as e:
            log.info("Error deleting registry value: %s", e)
            return

        try:
            log.info("Adding capegroup value: capegroup = %s", servicename)
            with CreateKeyEx(HKEY_LOCAL_MACHINE, svchost_path, 0, KEY_ALL_ACCESS | KEY_WOW64_64KEY) as key:
                SetValueEx(key, "capegroup", 0, REG_MULTI_SZ, [servicename])
        except Exception as e:
            log.info("Error setting registry value: %s", e)
            return

@disable_wow64_redirection
def start(self, path):
    ...
```

```python
...

@disable_wow64_redirection
def move_curdir(self, filepath):
    ...
```

```python
import os
import sys
import ctypes
from ctypes import wintypes, windll, POINTER
from ctypes.wintypes import BOOL, HANDLE
from functools import wraps


def disable_wow64_redirection(func):
    if os.name == 'nt' and sys.maxsize == 2**31 - 1:
        kernel32 = ctypes.windll.kernel32

        Wow64DisableWow64FsRedirection = kernel32.Wow64DisableWow64FsRedirection
        Wow64DisableWow64FsRedirection.restype = BOOL
        Wow64DisableWow64FsRedirection.argtypes = [POINTER(HANDLE)]

        Wow64RevertWow64FsRedirection = kernel32.Wow64RevertWow64FsRedirection
        Wow64RevertWow64FsRedirection.restype = BOOL
        Wow64RevertWow64FsRedirection.argtypes = [HANDLE]

    @wraps(func)
    def wrapper(*args, **kwargs):
        if os.name == 'nt' and sys.maxsize == 2**31 - 1:
            old_value = wintypes.HANDLE()
            Wow64DisableWow64FsRedirection(ctypes.byref(old_value))

            result = func(*args, **kwargs)

            Wow64RevertWow64FsRedirection(old_value)
            return result
        else:
            return func(*args, **kwargs)

    return wrapper
```

I also notice that `"%SystemRoot%\system32"` will not be interpret by python and will simply create a "%SystemRoot%" folder when calling `os.makedirs` and `os.path.exists` in `create_custom_folders`. We can simply change it to `C:\\Windows\\System32`.

https://github.com/kevoreilly/CAPEv2/blob/4d65838dfed81065c718bb95c68aaba029b06ae3/analyzer/windows/modules/packages/service_dll.py#L63

## capemon issues

After fixing all the issues related to WoW64, we can successfully run a 64-bit service DLL properly. However, this is not the end of the problem, as I still couldn't see the calc.exe pop up. I suspected this might be due to a capemon monitoring issue, and I discovered several problems in capemon.

First, when calling Process.inject, even though we can specify the injectmode, the parameter is not used at all in the function. This is also the case in the capemon loader, where __argv[5] is not used in the function. To fix this, I temporarily set DisableIATPatching to TRUE. In practice, we should properly handle the inject mode parameters.

https://github.com/kevoreilly/CAPEv2/blob/4d65838dfed81065c718bb95c68aaba029b06ae3/analyzer/windows/lib/api/process.py#L595

https://github.com/kevoreilly/capemon/blob/3d07de774acbfd5c3ff7d2c2d41dc8e7c1f0ceef/loader/loader/Loader.c#L1390

Next, I attached a debugger to svchost to determine where the program crashed. I found that it crashed in [capemon - CAPE/Debugger.c - CreateThreadBreakpoints](https://github.com/kevoreilly/capemon/blob/3d07de774acbfd5c3ff7d2c2d41dc8e7c1f0ceef/CAPE/Debugger.c#L145) and raised an `INVALID_HANDLE` error. After disabling it by modifying the `InitNewThreadBreakpoints` function to return FALSE at the beginning, it finally worked, and I could see the calc.exe pop up. But I don't know why this error happened.

![image](https://user-images.githubusercontent.com/18167730/230579076-db5abec6-3fda-4f8f-9a56-4344d53d95e2.png)

## Steps to Reproduce

https://capesandbox.com/analysis/378048/#

## Context

| Question         | Answer
|------------------|--------------------
| Git commit       |  c2600596ade8e3ec9ebbb960cc34a7e8da8d9489 (this is the latest commit I merge into my project, but I have some custom modification on the code)
| OS version       | Ubuntu 20.04, Windows 10

## Failure Logs

Please include any relevant log snippets or files here.

**Comment by kevoreilly** [comment]
Thank you for this detailed work - very much appreciated. I will work through these issues and get back to you!

**Comment by enzok** [comment]
For the INVALID_HANDLE error, you could use the INVALID_HANDLE_VALUE constant to check for an invalid handle before calling GetThreadId, and then GetLastError to check the error code that caused the handle to be invalid. As a workaround, instead of editing InitNewThreadBreakpoints, just skip the thread with the invalid handle.

**Comment by kevoreilly** [comment]
I will respond in more detail in time but cape's loader does not use the injectmode parameter, this is a remnant from cuckoo which will be removed from the analyzer in time. The loader is designed to pick the appropriate loading mechanism, in this case since the service process is already running the injection should be remote thread creation (neither queued APC or IAT patching are appropriate for existing processes) and this is always picked in the cases where there is no thread id supplied, Since we should not have a thread id for the pre-existing services.exe process, then thread creation should in theory be picked here - I will need to look into why this is not the case. However, I am on leave today so will not be able to look at this until tomorrow.

**Comment by kevoreilly** [comment]
Thanks @enzok for your reply - my plan is to solve the root problem rather than just work around it. Again, until I have time to look in detail, there isn't much point in saying too much other than there shouldn't be an invalid handle value stored there.

**Comment by oalieno** [comment]
Initially, I suspected that there might be an issue with the IAT injection method when spawning svchost.exe. However, after successfully executing my service DLL and testing it again, I discovered that the issue was not related to the IAT injection method. Therefore, there is no need to set DisableIATPatching or use any other approach.
About injecting services.exe, I guess it works correctly since I don't encounter issue with it.

**Comment by oalieno** [comment]
Thank you for the response, @enzok. I have updated the code as follows to check for `INVALID_HANDLE_VALUE` before calling `GetThreadId`. However, it still crashes in the same manner, leading me to believe that the `CurrentThreadBreakpoints->ThreadHandle` is neither 0 nor `INVALID_HANDLE_VALUE`. Further investigation will be required to resolve this issue.

```c
if (CurrentThreadBreakpoints->ThreadHandle && 
    CurrentThreadBreakpoints->ThreadHandle != INVALID_HANDLE_VALUE && 
    GetThreadId(CurrentThreadBreakpoints->ThreadHandle) == ThreadId)
```

**Comment by kevoreilly** [comment]
Unfortunately I have not been able to successfully recreate your wow64 redirection from these code snippets. Whilst I am grateful for the code you have supplied in the issue, in order to be able to work properly and efficiently it would be preferable for this type of work to be put into a pull request rather than an issue. That way there is no time wasted copying and pasting code and failing to get to the same codebase.

So I have created a pull request on your behalf and attempted to put together the code you have posted: https://github.com/kevoreilly/CAPEv2/pull/1480. It wasn't stated where the disable_wow64_redirection() function was actually implemented in file, so I picked lib/api/common/common.py.

Would you be able to check this pull request to get the code to the same state you have yours? Currently I am seeing a 32-bit process being created by services.exe suggesting the redirection disabling has not worked.

 ```
2023-04-11 11:03:45,940 [modules.packages.service_dll] INFO: Created service CAPEService (handle: 0x24072576)
2023-04-11 11:03:45,940 [modules.packages.service_dll] INFO: Adding Parameters value: System\CurrentControlSet\Services\CAPEService\Parameters -> ServiceDll = C:\Windows\system32\service_dll.dll
2023-04-11 11:03:45,940 [modules.packages.service_dll] INFO: Adding capegroup value: capegroup = CAPEService
...
2023-04-11 11:03:46,471 [root] INFO: Loaded monitor into process with pid 676
2023-04-11 11:03:46,487 [root] DEBUG: CreateProcessHandler: Injection info set for new process 7880: C:\Windows\SysWOW64\svchost.exe, ImageBase: 0x0000000000B20000
2023-04-11 11:03:46,487 [root] INFO: Announced 32-bit process name: svchost.exe pid: 7880
```

Thank you

**Comment by kevoreilly** [comment]
It looks like I have got it working just by fixing up a couple of missing imports...

**Comment by kevoreilly** [comment]
I've got the service dll running now but the call to `WTSQueryUserToken' appears to be failing...

```
CAPE Sandbox - Debugger log: Tue Apr 11 14:27:48 2023
Breakpoint 0 hit by instruction at 0x00007FFC3B081060 (thread 2576) RAX=0x7ffc3b081060 RCX=0x1 RDX=0x19e2ec1a928 RSI=0xffffffffffffffff RDI=0x19e2ec14af0 RSP=0x953e37fd98 *RSP=0x7ff744f94140 RBP=0x953e37fe69 R8=0x7ff744f99070 R9=0x8000000000000000 R10=0xfff8761020c R11=0xffffffffffffffff R12=0x19e2ec1a928 R13=0x19e2ec1a930 L"CAPEService" R15=0x7ffc3b081060 Xmm0.Low=0x31120000 Xmm0.High=0x32120000000000 Xmm1.Low=0x1d4003212000000
Break at 0x00007FFC3B081060 in service_dll.dll::ServiceMain (RVA 0x1060, thread 2576, ImageBase 0x00007FFC3B080000)
0x00007FFC3B081060  4057                     PUSH      RDI                            RSP=0x953e37fd90 *RSP=0x19e2ec14af0
0x00007FFC3B081062  4881EC00010000           SUB       RSP, 0x100                     RSP=0x953e37fc90 *RSP=0x953e37fe69
0x00007FFC3B081069  488B05982F0000           MOV       RAX, [RIP+0x2f98]              RAX=0x3e253715383c
0x00007FFC3B081070  4833C4                   XOR       RAX, RSP                       RAX=0x3eb00922c4ac
0x00007FFC3B081073  48898424F0000000         MOV       [RSP+0xf0], RAX               
0x00007FFC3B08107B  488D157EFFFFFF           LEA       RDX, [RIP-0x82]                RDX=0x7ffc3b081000
0x00007FFC3B081082  488D0DA7210000           LEA       RCX, [RIP+0x21a7]              RCX=0x7ffc3b083230 L"EvilSvc"
0x00007FFC3B081089  FF15811F0000             CALL      RegisterServiceCtrlHandlerW    RAX=0x19e2ec15c80 RCX=0x953e37fc70 RDX=0x7ffc3b083230 L"EvilSvc" R8=0 R9=0
0x00007FFC3B08108F  33FF                     XOR       EDI, EDI                       RDI=0
0x00007FFC3B081091  C705A535000020000000     MOV       DWORD [RIP+0x35a5], 0x20      
0x00007FFC3B08109B  488BC8                   MOV       RCX, RAX                       RCX=0x19e2ec15c80
0x00007FFC3B08109E  48890593350000           MOV       [RIP+0x3593], RAX             
0x00007FFC3B0810A5  488D1594350000           LEA       RDX, [RIP+0x3594]              RDX=0x7ffc3b084640
0x00007FFC3B0810AC  893D9E350000             MOV       [RIP+0x359e], EDI             
0x00007FFC3B0810B2  C7058835000002000000     MOV       DWORD [RIP+0x3588], 0x2       
0x00007FFC3B0810BC  FF15461F0000             CALL      SetServiceStatus               RAX=0x1 RCX=0x405e501c736e0000 RDX=0xa10 R8=0x19e2ec057d0 R9=0x953e37f608 R10=0xfff89276ccc R11=0x1000 Xmm0.Low=0 Xmm0.High=0 Xmm1.Low=0
0x00007FFC3B0810C2  4533C9                   XOR       R9D, R9D                       R9=0
0x00007FFC3B0810C5  8D5701                   LEA       EDX, [RDI+0x1]                 RDX=0x1
0x00007FFC3B0810C8  4533C0                   XOR       R8D, R8D                       R8=0
0x00007FFC3B0810CB  33C9                     XOR       ECX, ECX                       RCX=0
0x00007FFC3B0810CD  FF15751F0000             CALL      CreateEventW                   RAX=0x1f0 RCX=0x48cbb439dd400000 RDX=0 R8=0x953e37fb78 R9=0x953e37fc38 R10=0 R11=0x953e37fc80
0x00007FFC3B0810D3  488B0D5E350000           MOV       RCX, [RIP+0x355e]              RCX=0x19e2ec15c80
0x00007FFC3B0810DA  488D155F350000           LEA       RDX, [RIP+0x355f]              RDX=0x7ffc3b084640
0x00007FFC3B0810E1  48890548350000           MOV       [RIP+0x3548], RAX             
0x00007FFC3B0810E8  C7055235000004000000     MOV       DWORD [RIP+0x3552], 0x4       
0x00007FFC3B0810F2  FF15101F0000             CALL      SetServiceStatus               RAX=0x1 RCX=0x405e501c736e0000 RDX=0xa10 R8=0x19e2ec057d0 R9=0x953e37f608 R10=0xfff89276ccc R11=0x1000
0x00007FFC3B0810F8  0F57C0                   XORPS     XMM0, XMM0                    
0x00007FFC3B0810FB  C784248000000068000000   MOV       DWORD [RSP+0x80], 0x68        
0x00007FFC3B081106  33C0                     XOR       EAX, EAX                       RAX=0
0x00007FFC3B081108  0F11842484000000         MOVUPS    [RSP+0x84], XMM0              
0x00007FFC3B081110  898424E4000000           MOV       [RSP+0xe4], EAX               
0x00007FFC3B081117  0F11842494000000         MOVUPS    [RSP+0x94], XMM0              
0x00007FFC3B08111F  4889442478               MOV       [RSP+0x78], RAX               
0x00007FFC3B081124  0F118424A4000000         MOVUPS    [RSP+0xa4], XMM0              
0x00007FFC3B08112C  0F118424B4000000         MOVUPS    [RSP+0xb4], XMM0              
0x00007FFC3B081134  0F118424C4000000         MOVUPS    [RSP+0xc4], XMM0              
0x00007FFC3B08113C  0F118424D4000000         MOVUPS    [RSP+0xd4], XMM0              
0x00007FFC3B081144  0F11442468               MOVUPS    [RSP+0x68], XMM0              
0x00007FFC3B081149  FF15D11E0000             CALL      WTSGetActiveConsoleSessionId   RAX=0x3
0x00007FFC3B08114F  83F8FF                   CMP       EAX, -0x1                     
0x00007FFC3B081152  747D                     JZ        0x7f                          
0x00007FFC3B081154  488D542460               LEA       RDX, [RSP+0x60]                RDX=0x953e37fcf0
0x00007FFC3B081159  48897C2460               MOV       [RSP+0x60], RDI               
0x00007FFC3B08115E  8BC8                     MOV       ECX, EAX                       RCX=0x3
0x00007FFC3B081160  FF158A1F0000             CALL      QueryUserToken                
Exception 0x800703f0 at 0x00007FFC48184F69 in KERNELBASE.dll (RVA 0x34f69, thread 2576), flags 0x1 RAX=0 RCX=0x5d158db13dce0000 RDX=0 R8=0 R9=0x7f67095eb2c3 R10=0 R11=0x200 Xmm0.Low=0x7ffc4951f0a0 Xmm0.High=0x19e2ed02fa0
0x00007FFC3B081166  85C0                     TEST      EAX, EAX                      
0x00007FFC3B081168  7467                     JZ        0x69                          
0x00007FFC3B0811D1  488B0D58340000           MOV       RCX, [RIP+0x3458]              RCX=0x1f0
0x00007FFC3B0811D8  BAFFFFFFFF               MOV       EDX, 0xffffffff                RDX=0xffffffff
0x00007FFC3B0811DD  FF15451E0000             CALL      WaitForSingleObject 
```

**Comment by kevoreilly** [comment]
Right I've pushed a test monitor to the PR branch `service_dll` along with a few small code changes - everything now seems to be working other than the failing `QueryUserToken` call from the compiled sample dll.

Please could you test this branch to see how it fares on your instance.

**Comment by oalieno** [comment]
Apologies for the delayed response. I've tested your pull request, and it works on my instance as well. However, this is only a quick fix for 64-bit service DLLs. Is it acceptable to ignore 32-bit service DLLs, or should we change the submission option to service_dll(x64)?

**Comment by kevoreilly** [comment]
The big question is whether it is possible to run a 32-bit service dll on a 64-bit operating system. I was of the impression that it is not, but I cannot seem to find any definitive evidence either way...

Once this is established, we can test for bitness of the dll within the package and either handle each separately, or if it's not possible to run the dll, return an appropriate error.

**Comment by kevoreilly** [comment]
These changes were merged in https://github.com/kevoreilly/CAPEv2/commit/6437b84965ee0ca5d9918978a78b332524290105 - thanks for this contribution.

I haven't come across any evidence of 32-bit service dlls running on 64-bit Windows - we can cross that bridge when we come to it.

**Comment by joaogodinho** [comment]
Hi @kevoreilly, I was trying to run a `service_dll` and noticed that `disable_wow64_redirection` does not seem to be implemented: [/analyzer/windows/lib/common/common.py](/kevoreilly/CAPEv2/blob/master/analyzer/windows/lib/common/common.py), although but is imported from `service_dll.py`.

**Comment by kevoreilly** [comment]
Fix is in staging
